src/credentials.rs

```diff
@@ -31,55 +31,110 @@ pub struct Credentials {
     pub registry_tokens: HashMap<RegistryUri, String>,
 }
 
-impl Credentials {
-    fn location() -> eyre::Result<PathBuf> {
-        let home = home::home_dir().wrap_err("Failed to locate home directory")?;
+/// Error produced when locating the credentials file
+#[derive(Error, Debug)]
+pub enum LocateError {
+    /// The home folder where the credentials file is stored could not be determined.
+    ///
+    /// This implies that BUFFRS_HOME is not set and that the user's home directory
+    /// could not be detected.
+    #[error("BUFFRS_HOME unset and could not resolve user's home directory")]
+    MissingHome,
+}
 
-        let home = env::var("BUFFRS_HOME").map(PathBuf::from).unwrap_or(home);
+/// Error produced when checking if the credentials file exists
+#[derive(Error, Debug)]
+pub enum ExistsError {
+    /// Could not locate the credentials file
+    #[error("Failed to determine credentials location. {0}")]
+    Locate(LocateError),
+    /// Could not access the filesystem
+    #[error("IO error: {0}")]
+    Io(std::io::Error),
+}
+
+/// Error produced when loading the credentials file
+#[derive(Error, Debug)]
+pub enum ReadError {
+    /// Could not locate the credentials file
+    #[error("Failed to determine credentials location. {0}")]
+    Locate(LocateError),
+    /// Could not read the file from the filesystem
+    #[error("IO error: {0}")]
+    Io(std::io::Error),
+    /// Could not parse the TOML credentials data
+    #[error("Failed to deserialize credentials. Cause: {0}")]
+    Toml(toml::de::Error),
+}
+
+/// Error produced when updating the credentials file
+#[derive(Error, Debug)]
+pub enum WriteError {
```

Comment:
Which ones? All `WriteError` structs? There are slight differences, but maybe we can take a union of fields.

Reply:
`WriteError` `ReadError` etc – imo they can all be joined to either one `CredentialsError` or even something like `FileError` which could be reused across the manifest, credentials, config impls etc. I dont have a strong opinion on what is better, but i think having one error enum per function is to granular

Reply:
See [my recent comment](https://github.com/helsing-ai/buffrs/pull/95#issuecomment-1741741985) on why I think this kind of merging is a bad idea. We could also have a wrapper type at the module level but I think that is overkill. I may still do some merging across variants, but I'm now skeptical about merging errors altogether.

In fact there may be some benefit to defining unique types even if there's exactly one failure mode, in order to support future extensibility in case the implementation becomes more complex (more specifically, functions like `Lockfile::exists()` that currently return `IoError` would return their own local type that works like `IoError` but is marked as `#[non_exhaustive]` so that it could be extended without a break). Not 100% on this yet but that's my inclination.